components/encryption/src/io.rs

```diff
@@ -334,7 +340,10 @@ impl<W: Write> Write for CrypterWriter<W> {
         if let Some(crypter) = self.crypter.as_mut() {
             let crypted = crypter.do_crypter(buf)?;
             debug_assert!(crypted.len() == buf.len());
-            self.writer.write(crypted)
+            let r = self.writer.write(crypted);
+            let missing = buf.len() - r.as_ref().unwrap_or(&0);
+            crypter.lazy_reset_crypter(crypter.offset - missing as u64);
```

Comment:
A note here: Will we suffer really bad performance when the whole stream is long and the underlying writer writes a little bytes in each write? 

If we encrypt the whole buffer here each time, we may need to run AES-256 in `O(n^2)` bytes (if the openssl implementation doesn't cache the keystream), or do XOR with these bytes, at least.

But anyway this won't make things worse, and I have no idea about how to improve this.

Reply:
Sure, but it should be rare. This could be improved by calling `write_all` instead of `write`. But it would be a violation of `Write` interface, and if there's an non-interrupt error, it causes the underlying file to enter "uncertain offset" state but we have not way to alert the user through the outer `write` interface.